### PR TITLE
Adding Hinge Loss Layer

### PR DESCRIPTION
**Pull request by abster12**
I have just added the files for the model. could you help me progress from here and also if i have to add or delete some files from the current status

### CODE REVIEW

**Comment by rbharath** [comment]
@abster12 Thanks for getting this started! Here are a couple tips:

- The goal is to implement the SCScore model in `TensorGraph`. Check the tutorials at deepchem.io for an introduction to TensorGraph
- Before that, I'd recommend making sure that you can use the original code to duplicate the results claimed by the implementors
- Then, try making a TensorGraph implementation and comparing against the original implementation.

**Comment by lilleswing** [comment]
Verifying reproduction of the original paper is tricky here (I cost out the full original dataset to $10,000).  However you can train on the smaller datasets provided in the git repo and versify internal consistency.

We have all the layers needed in TensorGraph besides Hinge Loss.

**Comment by abster12** [comment]
there are 3 different models in the repository the 1024bool,1024uint8 and 2048 bool. What is the main difference between them?

**Comment by lilleswing** [comment]
Those are three different featurization schemes as discussed in the paper.  According to the paper they had equivelent results so the auther suggests the cheapest one (1024 bit chiral fingerprints)

**Comment by abster12** [comment]
Can anyone help me with the architecture of the model?

**Comment by abster12** [comment]
I was having problem with the test and so it was not written properly

File: deepchem/models/tensorgraph/layers.py [hunk 1]
```diff
@@ -7,7 +7,7 @@
 import tensorflow as tf
 import numpy as np
 
-from deepchem.models.tensorgraph import model_ops, initializations, regularizers, activations
+from deepchem.nn import model_ops, initializations, regularizers, activations
```
Comment (rbharath) [inline deepchem/models/tensorgraph/layers.py]: deepchem.nn has been removed recently so you might be on an old version of DeepChem

**Comment by abster12** [comment]
should I just rename this PR and work on the scscore model on a new PR

**Comment by rbharath** [comment]
Yes, that sounds like a good idea. Renaming to "Adding Hinge Loss Layer" would work well.

File: deepchem/models/tensorgraph/layers.py [hunk 2]
```diff
@@ -4303,3 +4303,29 @@ def batch_mat_mult(self, A, B):
     result = tf.matmul(A_reshape, B)
     result = tf.reshape(result, tf.stack([A_shape[0], A_shape[1], axis_2]))
     return result
+
+
+class Hingeloss():
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: Class should inherit layer

File: deepchem/models/tensorgraph/layers.py [hunk 3]
```diff
@@ -4303,3 +4303,29 @@ def batch_mat_mult(self, A, B):
...
+               **kwargs):
+
+    self.weights = weights
+    self.batch_size = batch_size
+    self.labels = labels
+    self.reg = reg
+    super(Hingeloss, self).__init__(in_layers, **kwargs)
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    inputs = self.get_input_tensors(in_layers)
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: can you add documentation on the in layers?  It seems like batch_size is a tensor and not a scaler.  I am confused at what each of them should be.

File: deepchem/models/tensorgraph/tests/test_layers.py [hunk 4]
```diff
@@ -875,3 +875,16 @@ def test_IRV(self):
...
+
+  def test_hingeloss(self):
+
+    batch_size = 10
+    reg = 0.00001
+    labels = np.array([0, 1])
+    weights = tf.Variable(10)
+    input_tensor = np.random.rand(weights, batch_size, labels, reg)
+    with self.test_session() as sess:
+      input_tensor = tf.convert_to_tensor(input_tensor, dtype=tf.float32)
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/tests/test_layers.py]: The inputs/outputs described here are not the same as in the layer.

File: deepchem/models/tensorgraph/layers.py (hunk 3 above)
Comment (abster12) [reply deepchem/models/tensorgraph/layers.py]: I didn't get what you are trying to ask, what am I supposed to add documentation for/

Comment (abster12) [reply deepchem/models/tensorgraph/layers.py]: I think I got it. Will do that in a few minutes.

**Comment by abster12** [comment]
I have added the documentation so if you could see if that was what you meant

File: deepchem/models/tensorgraph/layers.py (hunk 3 above)
Comment (peastman) [reply deepchem/models/tensorgraph/layers.py]: Just add docstrings for the class and for `__init__()`.  We're trying to be good about having documentation for all layers, though some of the older ones are missing it.

Comment (abster12) [reply deepchem/models/tensorgraph/layers.py]: ok, I added for the __init__() in the update and also add about the class

File: deepchem/models/tensorgraph/layers.py [hunk 5]
```diff
@@ -4303,3 +4303,38 @@ def batch_mat_mult(self, A, B):
...
+    """
+
+    self.weights = weights
+    self.batch_size = batch_size
+    self.labels = labels
+    self.reg = reg
+    super(Hingeloss, self).__init__(in_layers, **kwargs)
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    inputs = self.get_input_tensors(in_layers)
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: Why are we not calling tensorflow's built in hinge loss.
https://www.tensorflow.org/api_docs/python/tf/losses/hinge_loss

Comment (abster12) [reply deepchem/models/tensorgraph/layers.py]: I thought we wanted to have our own, I will change that

File: deepchem/models/tensorgraph/layers.py [hunk 6]
```diff
@@ -4303,3 +4303,38 @@ def batch_mat_mult(self, A, B):
...
+    weights, batch_size, labels, reg = inputs[0], inputs[1], inputs[2], inputs[
+        3]
+    scores = tf.matmul(batch_size, weights)
+    labels_i = scores[np.arange(scores.shape[0]), labels]
+    loss_i = tf.maximum(0, scores - tf.transpose(labels_i) + 1)
+    loss = tf.reduce_mean(tf.sum(loss_i), axis=1)
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: tf.sum?  This isn't a function.  I believe you are looking for tf.reduce_sum.

File: deepchem/models/tensorgraph/tests/test_layers.py [hunk 7]
```diff
@@ -875,3 +875,16 @@ def test_IRV(self):
       assert out_tensor.shape == (batch_size, n_tasks)
       irv_reg = IRVRegularize(irv_layer, 1.)()
       assert irv_reg.eval() >= 0
+
+  def test_hingeloss(self):
```
Comment (rbharath) [inline deepchem/models/tensorgraph/tests/test_layers.py]: Could you also add a test that the new Hinge Loss layer can be pickled? (Should be added in https://github.com/deepchem/deepchem/blob/master/deepchem/models/tensorgraph/tests/test_layers_pickle.py with other tests for pickling)

File: deepchem/models/tensorgraph/layers.py [hunk 8]
```diff
@@ -4303,3 +4303,23 @@ def batch_mat_mult(self, A, B):
...
+
+  def __init__(self, in_layers=None, **kwargs):
+    super(Hingeloss, self).__init__(in_layers, **kwargs)
+    try:
+      self._shape = self.in_layers[1].shape
+    except:
+      pass
+
+  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
+    inputs = self._get_input_tensors(in_layers)
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: Please document that in_layers should be lables, logits.

Also is labels one hot encoded or not here?

File: deepchem/models/tensorgraph/layers.py [hunk 9]
```diff
@@ -4303,3 +4303,23 @@ def batch_mat_mult(self, A, B):
...
     result = tf.reshape(result, tf.stack([A_shape[0], A_shape[1], axis_2]))
     return result
+
+
+class Hingeloss(Layer):
+
+  def __init__(self, in_layers=None, **kwargs):
+    super(Hingeloss, self).__init__(in_layers, **kwargs)
+    try:
+      self._shape = self.in_layers[1].shape
```
Comment (lilleswing) [inline deepchem/models/tensorgraph/layers.py]: This shape is incorrect -- that is why your tests are failing.
https://www.tensorflow.org/api_docs/python/tf/losses/hinge_loss

The shape would be correct if you call
```python
tf.losses.hing_loss(labels=labels, logits=logits, reduction=None)
```
with the reduction=None keyword.

File: deepchem/models/tensorgraph/layers.py (hunk 8 above)
Comment (abster12) [reply deepchem/models/tensorgraph/layers.py]: yes the labels are one hot encoded here

File: deepchem/models/tensorgraph/layers.py (hunk 9 above)
Comment (abster12) [reply deepchem/models/tensorgraph/layers.py]: that was a doubt, so calling it with reduction=None is not a problem?

**Comment by abster12** [comment]
I will add the documentation of this layer alongwith some others in a different PR

**Comment by rbharath** [comment]
@abster12 Thanks for the fixes! Would you mind adding the documentation in this PR itself? We're trying to make sure that all new layers have proper documentation as they're added in.

**Comment by abster12** [comment]
No problem. I will add it in this PR itself.

**Comment by lilleswing** [comment]
This code Looks Good To Me!  Great Addition to the code @abster12 

Merging.
